// add/add.js
function visArray(array) {
  var viz = "";
  for (var i = 0; i < array.length; i++) {
    for (var j = 0; j < array[i].length; j++) {
    	if (array[i][j] === " ") {
    		viz = viz + "|  " + array[i][j] + " ";
    	} else if (array[i][j] === 10) {
    		viz = viz + "| " + array[i][j] + " ";
    	} else {
    		viz = viz + "|  " + array[i][j] + " ";
    	}
    }
    viz = viz + "| " + "\n";
  }
  return viz;
}

function user(movies) {
	var arr = [];
	for (var i = 0; i < movies; i++) {
		arr.push(Math.round(10 * Math.random()));
	}
	return arr;
}

function genUsers(users, movies) {
	var arr = [];
	for (var i = 0; i < users; i++) {
		arr.push(user(movies).slice());
	}
	return arr;
}

function findAverage(array) {
	if (array.length == 0) {
		return "Error"
	}
	var averages = [];
	for (var i = 0; i < array[0].length; i++) {
		var sum = 0;
		for (var j = 0; j < array.length; j++) {
			sum = sum + array[j][i];
		}
		averages.push(sum / array.length);
	}
	return averages;
}

function genIncompleteUsers(users, movies) {
	function incomplete(movies) {
		var arr = [];
		for (var i = 0; i < movies; i++) {
			var choice = Math.round(Math.random());
			if (choice == 0) {
				arr.push(" ");
			} else {
				arr.push(Math.round(10 * Math.random()));
			}
		}
		return arr;
	}
	var arr = [];
	for (var i = 0; i < users; i++) {
		arr.push(incomplete(movies));
	}
	return arr;
}

function isComplete(array) {
	for (var i = 0; i < array.length; i++) {
		for (var j = 0; j < array[i].length; j++) {
			if (array[i][j] === " ") {
				return false;
			}
		}
	}
	return true;
}

function findIncompleteAverages(array) {
	if (isComplete(array)) {
		return findAverage(array);
	}
	if (array.length == 0) {
		return "Error"
	}
	var averages = [];
	for (var i = 0; i < array[0].length; i++) {
		var sum = 0;
		var n = 0;
		for (var j = 0; j < array.length; j++) {
			if (!(array[j][i] === " ")) {
				sum = sum + array[j][i];
				n++;
			}
		}
		averages.push(sum / n);
	}
	return averages;
}

// DO NOT DELETE ANYTHING BELOW THIS LINE OR NOTHING WILL WORK

module.exports = {f1 : user, f2 : genUsers, f3 : findAverage, f4 : genIncompleteUsers, f5 : isComplete, f6 : findIncompleteAverages}
