// test/add.js
function stringCompare(string, text, startingIndex) {
  var n = string.length;
  if (startingIndex + n - 1 > text.length) {
    return false;
  }
  for (var i = 0; i < n; i++) {
    // compare characters in a string here
    if (text.charAt(startingIndex + i) !== string.charAt(i)) {
      return false;
    }
  }
  return true;
}

function searchString(string, text) {
  var n = string.length;
  var t = text.length;
  if (t < n) {
    return false;
  }
  // search text for the first character in string
  var count = 0;
  for (var i = 0; i < t; i++) {
    if (text.charAt(i) === string.charAt(0)) {
      if (stringCompare(string, text, i)){
        count++;
      }
    }
  }
  return count;
}

function minBS(array, item) {
	var l = 0;
	var r = array.length - 1;
	if (r == 0) {
		return false;
	}
	// complete the following code, which has been commented out (and uncomment it)
	// put in something for MISSING1, MISSING2 and MISSING3
	/*
	while (r >= l) {
		var m = Math.floor((l + r) / 2);
		if (MISSING1) {
			return MISSING2;
		} else if (MISSING3) {
			l = m + 1;
		} else {
			r = m - 1;
		}
	}
	*/
	return l;
}

function genP(n) {
	// this generates an array storing two-element arrays: each element stores an integer i and the Boolean true
	var array = [];
	for (var i = 2; i <= n; i++) {
		array.push([i,true]);
	}
	// complete the code here to implement the Sieve of Eratosthenes
	// if the value stored in array is not prime then the associated Boolean should be changed to false

	// this will loop over all elements in array and if the Boolean in each element is true, then it pushes that number to out
	// out will store all primes up to (and possibly including) n
	var out = [];
	for (var i = 0; i < array.length; i++){
		if (array[i][1]) {
			out.push(array[i][0]);
		}
	}
	return out;
}

function randomPrime(len) {
	//complete the code to return a random prime number of length len, i.e. with len digits

	//remember that the smallest number of length len is 10**(len-1), and the largest number of length len is 10**(len)-1
}

function minBinarySearch(array,item) {
	//this should return the index of the first element with a value larger than item
	var l = 0;
	var r = array.length - 1;
	if (r == 0) {
		return false;
	}
	while (r >= l) {
		var m = Math.floor((l + r) / 2);
		if (array[m] === item) {
			return m + 1;
		} else if (array[m] < item) {
			l = m + 1;
		} else {
			r = m - 1;
		}
	}
	return l;
}

function genPrimes(n) {
	var array = [];
	for (var i = 2; i <= n; i++) {
		array.push([i,true]);
	}
	for (var i = 2; i <= Math.ceil(Math.sqrt(n)); i++) {
		if (array[i-2][1]) {
			for (var j = 2; j <= n / i; j++) {
				array[(i*j) - 2][1] = false;
			}
		}
	}
	var out = [];
	for (var i = 0; i < array.length; i++){
		if (array[i][1]) {
			out.push(array[i][0]);
		}
	}
	return out;
}

module.exports = {genP : genPrimes, minBS : minBinarySearch}
